Commands added for supporting IPv4 MPLS Layer 3 VPN.

KaushikNiral opened the pull request:
1. New address-family ipv4-vpn added.
2. New neighbor command added for address-family ipv4-vpn.
3. New label command added inside address-family ipv4-unicast.
4. Json script for all the above commands are handled.

KaushikNiral committed:
Commands added for supporting IPv4 MPLS Layer 3 VPN.
1. New address-family ipv4-vpn added.
2. New neighbor command added for address-family ipv4-vpn.
3. New label command added inside address-family ipv4-unicast.
4. Json script for all the above commands are handled.

Signed-off-by: kaushik <kaushik@niralnetworks.com>
Signed-off-by: harios <hari@niralnetworks.com>

deastoe commented on yang/vyatta-protocols-frr-bgp-v1.yang:
```diff
@@ -1291,6 +1352,95 @@ module vyatta-protocols-frr-bgp-v1 {
 						uses neigh_peer_grp_ipv4_lists;
 					}
+
+          container ipv4-vpn {
```
This container is under the peer-group configuration rather than the neighbor configuration, but the command translations above are looking for neighbor configuration.

So I'm not sure these diffs are working right now? If you want per-neighbor configuration then I think this should be under the bgp-params-neighbor grouping.

KaushikNiral replied:
Sorry, actually the container ipv4-vpn is placed under the bgp-params-neighbor grouping but somehow it went in wrong place in this PR.
It is fixed now, will send it for review shortly. 
I have attached a screenshot verifying that. 
![Saving-FRR_Config_file](https://user-images.githubusercontent.com/62253226/79356131-3324f400-7f5c-11ea-8513-4a206af9d54c.png)

deastoe commented on yang/vyatta-protocols-frr-bgp-v1.yang:
```diff
@@ -68,6 +68,32 @@ module vyatta-protocols-frr-bgp-v1 {
 		description "Inital version.";
 	}
 
+  typedef rt-rd {
```
This file uses Tab characters for indentation. Please could you use the same for your additions?

KaushikNiral replied:
We tried to fix this issue. 
We will send for review shortly.

KaushikNiral committed:
Commands added for supporting IPv4 MPLS Layer 3 VPN.
1. The container ipv4-vpn added under bgp-params-neighbor grouping.
2. The issue with indentation is resolved.

Signed-off-by: harios <hari@niralnetworks.com>

wivory commented on yang/vyatta-protocols-frr-bgp-v1.yang:
```diff
@@ -68,6 +68,32 @@ module vyatta-protocols-frr-bgp-v1 {
 		description "Inital version.";
 	}
 
+	typedef rt-rd {
```
Any change to a YANG file requires that you add a new revision statement for the change.

KaushikNiral replied:
Hi Wivory, 
This we have resolved.

wivory replied:
You've not tested this have you?  Our YANG compiler will refuse to accept this as the revision statements are not in reverse chronological order.  The 'pyang' tool will also object.

NB - if you have tested this, and your image boots and is configurable with this YANG, please let me know as it certainly shouldn't be working!

KaushikNiral replied:
Hi Wivory, 

We have tested it but we replaced the old revision statement with new one so we didn't see this issue. 
We have added the new revision statement in the correct order and tested.
This changes are also committed. 

Thanks,

wivory commented on yang/vyatta-protocols-frr-bgp-v1.yang:
```diff
@@ -1521,6 +1583,95 @@ module vyatta-protocols-frr-bgp-v1 {
+						uses peer-grp-neighbor-afi-common-settings {
+							refine filter-list/export {
+								must "not(../../peer-group)" {
+									error-message "You may not configure filter-list export for a neighbor in peer-group\n";
```
IETF max line length is 73 characters, and in DANOS we generally stick to under 80 characters for YANG files so you don't end up having to scroll sideways to read some lines (which really slows down reviews).  You also lose the context of the surrounding lines if they disappear off the left of the screen.

Furthermore, in this case you have a set of multiple changes close enough together that the side-scroll bar is miles down the page, making it incredibly difficult to scroll sideways.  I have to page down, scroll sideways, then page back up again.

KaushikNiral replied:
We understand the issue. 
We tried to be consistent with the existing file, so we did it in that manner. 

Now we have tried to rectify line length our part of code. Please have a look and suggest.

wivory commented on yang/vyatta-protocols-frr-bgp-v1.yang:
```diff
@@ -1521,6 +1583,95 @@ module vyatta-protocols-frr-bgp-v1 {
+								}
+							}
+						}
+						uses neigh_peer_grp_ipv4_lists {
```
For future reference, please always use hyphens in names, never underscores.

KaushikNiral replied:
Okay we will follow that. 
But here the grouping is already mentioned in this way so we just used in our container ipv4-vpn.

Thanks.

wivory replied:
Wasn't asking you to change *existing* code - that's unnecessary churn.  Just making a recommendation for naming in future as a matter of style / consistency.

KaushikNiral replied:
Okay got it. 

Thanks,

dewi-morgan commented:
Can you please detail the testing you have performed and which topologies ?

Are you testing vrf leaking or bgp mpls l3vpn ? If the later, I don't see the frr **label vpn export auto** under address-family ipv4-unicast being used, only static label value; is there a reason for not providing auto ?

If you have an l3vpn topology (ce-pe-p-pe-ce), can you provide additional detail e.g. the output of **sudo ip route show vrfVRFNAME** and **sh dataplane route routing-instance VRFNAME** for the bgp label stack being used on each PE router ? Also the corresponding ILM table  **sudo ip -f mpls route show** and **sh dataplane mpls label-table** on each PE router ?

Thank you,

hariosniral committed:
Commands added for supporting IPv4 MPLS Layer 3 VPN.
1. Revision added.
2. Rectified indentation.

Signed-off-by: harios <hari@niralnetworks.com>

KaushikNiral commented:
> Can you please detail the testing you have performed and which topologies ?
> 
> Are you testing vrf leaking or bgp mpls l3vpn ? If the later, I don't see the frr **label vpn export auto** under address-family ipv4-unicast being used, only static label value; is there a reason for not providing auto ?
> 
> If you have an l3vpn topology (ce-pe-p-pe-ce), can you provide additional detail e.g. the output of **sudo ip route show vrfVRFNAME** and **sh dataplane route routing-instance VRFNAME** for the bgp label stack being used on each PE router ? Also the corresponding ILM table **sudo ip -f mpls route show** and **sh dataplane mpls label-table** on each PE router ?
> 
> Thank you,

Hi @dewi-morgan , 

Sure, we will share you the details of our topology. 
We have performed the IPv4 BGP MPLS L3VPN test. We are initially trying to do it with static labels so we added **label vpn export <label-value>** under address-family ipv4-unicast where label value  ranges from <0-1048575>.
In the coming days we will also add  **label vpn export auto** under address-family ipv4-unicast. 
We are also working in the vyatta-dataplane for making the IPv4 BGP MPLS L3VPN functional end to end. 
FYI we have tested this feature with FRR version 7.3 
Please find all the screenshot attached below. 
1. The Topology (ce-pe-p-pe-ce)   
    R1 and R5 are CE (gns3)
    R2 is PE (gns3)
    R3 is P  (gns3)
    cloud 1 is PE ( DANOS )
![setup](https://user-images.githubusercontent.com/62253226/79455654-d7657400-800a-11ea-96fc-030882ac774f.png)

2. Display of **sudo ip route show vrfVRFNAME** 
![ip_route_routing_instance_table](https://user-images.githubusercontent.com/62253226/79455870-29a69500-800b-11ea-9fc5-dfadf02f3a48.png)

3. Display of **sh dataplane route routing-instance VRFNAME**
![dataplane_route_routing_instance_table](https://user-images.githubusercontent.com/62253226/79455944-48a52700-800b-11ea-85d9-24506b6d0c14.png)

4. Display of ILM table **sudo ip -f mpls route show**
![ip-f_mpls_route_table](https://user-images.githubusercontent.com/62253226/79456049-71c5b780-800b-11ea-9f29-7b8ada659b25.png)

5. Display of **sh dataplane mpls label-table**
![dataplane_mpls_label-table](https://user-images.githubusercontent.com/62253226/79456113-8efa8600-800b-11ea-9a95-c35df22a0d4d.png)

The dataplane mpls label-table is not populated so we are debugging this in vyatta-dataplane.

Thanks,

hariosniral committed:
Commands added for supporting IPv4 MPLS Layer 3 VPN.
1 Added correctly the new revision statement for the changes.

Signed-off-by: harios <hari@niralnetworks.com>

wivory reviewed (approved):
YANG changes approved.

dewi-morgan commented on yang/vyatta-protocols-frr-bgp-v1.yang:
```diff
@@ -1776,32 +1777,34 @@ module vyatta-protocols-frr-bgp-v1 {
-          configd:help "BGP IPv4 VPN settings";
-          uses network-ipv4-vpn;
-        }
+				container ipv4-vpn {
```
Rather than use container ipv4-vpn, can this be changed to vpnv4-unicast ? I understand that the ipv4-vpn is used by the underlying frr configuration, but the vpnv4-unicast / vpnv6-unicast  naming is in line with other vendors usage. The mapping would then be from vpnv4-unicast to ipv4-vpn to generate the corresponding frr config from the yang.

KaushikNiral replied:
> Rather than use container ipv4-vpn, can this be changed to vpnv4-unicast ? I understand that the ipv4-vpn is used by the underlying frr configuration, but the vpnv4-unicast / vpnv6-unicast naming is in line with other vendors usage. The mapping would then be from vpnv4-unicast to ipv4-vpn to generate the corresponding frr config from the yang.

Hi , 

We took a note for the change you mentioned here. 
When we send PR on the commands for supporting 6PE and 6VPE feature we will rectify that. 

Thanks, 
Kaushik

dewi-morgan replied:
looks good, Thank You,

hariosniral committed:
Commands added for supporting 6VPE and 6PE features.

1. Added "vpnv6-unicast" and "ipv6-labeled-unicast" address family in bgp.
2. Added support for "label vpn export auto" in bgp.
3. The address-family "ipv4-vpn" is now replaced with "vpnv4-unicast".

Signed-off-by: harios <hari@niralnetworks.com>

deastoe commented:
Hi @KaushikNiral, @hariosniral,

There are now some conflicts in these diffs. Please could you do a rebase and squash, then we can finish the review and get this merged.

To resolve the revision statement conflicts, please simply update the description on the existing "2020-11-23" revision.

Thanks.

dewi-morgan commented on yang/vyatta-protocols-frr-bgp-v1.yang:
```diff
@@ -216,6 +242,41 @@ module vyatta-protocols-frr-bgp-v1 {
+        type rt-rd;
+        default "100:1";
+      }
+      leaf label {
```
why is this needed ? bgp should allocate the vpn label. The "label vpn export auto" configuration assigns a vrf label ? Why can't this be done instead of using a static label ?

dewi-morgan commented on yang/vyatta-protocols-frr-bgp-v1.yang:
```diff
@@ -1626,10 +1776,33 @@ module vyatta-protocols-frr-bgp-v1 {
+              leaf export {
+                description "For routes leaked from current address-family to vpn";
+                configd:help "For routes leaked from current address-family to vpn";
+                type uint32 {
```
why is this needed ? bgp should allocate the vpn label. The "label vpn export auto" configuration assigns a vrf label ? Why can't this be done instead of using a static label ?

dewi-morgan commented on yang/vyatta-protocols-frr-bgp-v1.yang:
```diff
@@ -216,6 +242,41 @@ module vyatta-protocols-frr-bgp-v1 {
+      leaf addr {
+        description "BGP network";
+        type types:ipv4-prefix {
+          configd:normalize "normalize ipv4-prefix";
```
why is this required from ? Doesn't it make more sense to specify the network from the routing instance for the address family  ? Which rd is used if you configure e.g. 

set protocols bgp 65201 address-family ipv4-vpn network 1.2.3.0/24

dewi-morgan commented on yang/vyatta-protocols-frr-bgp-v1.yang:
```diff
@@ -321,6 +390,40 @@ module vyatta-protocols-frr-bgp-v1 {
+				type rt-rd;
+				default "100:1";
+			}
+			leaf label {
```
setting a static vpn label should not be necessary. I don't think this should be exposed in the yang.

dewi-morgan commented on yang/vyatta-protocols-frr-bgp-v1.yang:
```diff
@@ -1626,10 +2042,41 @@ module vyatta-protocols-frr-bgp-v1 {
+								description "For routes leaked from current address-family to vpn";
+								configd:help "For routes leaked from current address-family to vpn";
+								type union {
+									type uint32 {
```
I don't think setting a static vpn label is necessary. I don't think this should be exposed in the yang.

dewi-morgan commented on yang/vyatta-protocols-frr-bgp-v1.yang:
```diff
@@ -1641,10 +2088,39 @@ module vyatta-protocols-frr-bgp-v1 {
+								configd:help "For routes leaked from current address-family to vpn";
+								type union {
+									type uint32 {
+										range 0..1048575;
```
same comment, I don't think setting a static bgp label should be exposed in the yang.

dewi-morgan reviewed (changes requested):
I think bgp should be allocating labels rather than asking the user to select a static label. We should now allow the user to select static bgp vpn labels in the yang. Setting "label vpn export auto" allocates and programs an ILM vpn bgp label. I think this configuration can be applied via the configs/commands/bgp_vrf.json instead e.g.

"/routing/routing-instance/@element/protocols/bgp/@element/address-family/ipv4-unicast/route-distinguisher": ["rd vpn export {/@text}", "import vpn", "export vpn", "label vpn export auto"],

ekoyle commented:
What is left for this issue?  Is it only hung up on the static labels?
